crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1648,57 +1599,10 @@ impl Client {
         let request = assign!(join_room_by_id_or_alias::v3::Request::new(alias), {
             server_name: server_names,
         });
-
-        let (tx, mut rx) = mpsc::channel::<room::Room>(1);
-
-        let user_id = self.user_id().ok_or(Error::AuthenticationRequired)?.to_owned();
-
-        let now = MilliSecondsSinceUnixEpoch::now();
-        let handle = self.add_event_handler({
-            move |event: SyncStateEvent<RoomMemberEventContent>, room: room::Room| {
-                let mut tx = tx.clone();
-                let user_id = user_id.clone();
-
-                async move {
-                    if *event.membership() == MembershipState::Join
-                        && *event.state_key() == user_id
-                        && event.origin_server_ts() > now
-                    {
-                        if let Err(e) = tx.send(room).await {
-                            debug!(
-                                "Sending event from event_handler failed, \
-                                 receiver not ready: {}",
-                                e
-                            );
-                        }
-                    }
-                }
-            }
-        });
-
-        let _guard = self.event_handler_drop_guard(handle);
-
         let response = self.send(request, None).await?;
-        let room_id = response.room_id;
-
-        loop {
-            let room = StreamExt::next(&mut rx)
-                .await
-                .expect("receive joined room result from event handler");
-
-            if room.room_id() != room_id {
-                continue;
-            }
-
-            debug!("received RoomMemberEvent corresponding to requested join");
-
-            if let room::Room::Joined(joined_room) = room {
-                return Ok(joined_room);
-            } else {
-                warn!("Corresponding Room not in state: joined");
-                return Err(Error::InconsistentState);
-            };
-        }
+        let base_room =
+            self.base_client().get_or_create_room(&response.room_id, RoomType::Joined).await;
+        Ok(room::Joined::new(self, base_room).unwrap())
```

Comment:
This can panic.

`get_or_create_room` will only use the given `RoomType` if there isn't a `Room` already in the store, if we're rejoining a `Room` there will be a left room in the store.

So we get a `base_room` that is in the `RoomType::Left` state, now we try to create a `room::Joined` from that `base_room`.

Critically `Joined::new()` checks the `Room::room_type()` and if it doesn't match it returns `None`.

I think this panics every time you rejoin a room.